components/script/dom/node.rs

```diff
@@ -2726,18 +2779,109 @@ impl NodeMethods for Node {
     }
 
     // https://dom.spec.whatwg.org/#dom-node-comparedocumentposition
-    fn CompareDocumentPosition(&self, other: &Node) -> u16 {
+    fn CompareDocumentPosition(&self, other: Option<&Node>) -> u16 {
         // step 1.
-        if self == other {
+        if Some(self) == other {
             return 0;
         }
 
-        // FIXME(emilio): This will eventually need to handle attribute nodes.
+        // step 2
+        let node1: Option<&Node> = other;
+        let node2: Option<&Node> = Some(self);
 
-        let mut self_and_ancestors = self
+        // step 3
+        let attr1: Option<&Attr>;
+        let attr2: Option<&Attr>;
```

Comment:
Just get rid of these and have `let`s below, leaving a `step 3: nop` comment.

Or, make this `let mut attr1 = None`.

Reply:
Yeah I think it's better if you `let mut attr1 = None` and then redo the code below to closer match the spec text's structure. So step 4 would be

```rust
if let Some(a) = node2.unwrap().downcast::<Attr>() {
    attr2 = Some(a);
    node2 = a.GetOwnerElement().map(|e| e.upcast());
}
```

Step 5 would be similar, with an additional subpart for 5.1

Reply:
That should also fix any lifetime issues